src/engine/client/text.cpp

```diff
@@ -1775,34 +1725,79 @@ class CTextRender : public IEngineTextRender
 				}
 
 				pBitmap = &pFont->m_FtFace->glyph->bitmap; // ignore_convention
-
-				int MaxSizeWidth = (MaxWidth - WidthLastChars);
-				if(MaxSizeWidth > 0)
-				{
-					int SlotW = ((((unsigned int)MaxSizeWidth) < pBitmap->width) ? MaxSizeWidth : pBitmap->width);
-					int SlotH = pBitmap->rows;
-					int SlotSize = SlotW*SlotH;
-
-					// prepare glyph data
-					mem_zero(ms_aGlyphData, SlotSize);
-
-					if(pBitmap->pixel_mode == FT_PIXEL_MODE_GRAY) // ignore_convention
+				
+				int SlotW = pBitmap->width;
+				int SlotH = pBitmap->rows;
+				int SlotSize = SlotW*SlotH;
+				
+				// prepare glyph data
+				mem_zero(ms_aGlyphData, SlotSize);
+
+				if(pBitmap->pixel_mode == FT_PIXEL_MODE_GRAY) // ignore_convention
 					{
 						for(py = 0; py < (unsigned)SlotH; py++) // ignore_convention
 							for(px = 0; px < (unsigned)SlotW; px++)
 							{
 								ms_aGlyphData[(py)*SlotW + px] = pBitmap->buffer[py*pBitmap->width + px]; // ignore_convention
 							}
 					}
+				
+				Graphics()->LoadTextureRawSub(TextureID, x + WidthLastChars, y, SlotW, SlotH, CImageInfo::FORMAT_ALPHA, ms_aGlyphData);
+				WidthLastChars += (SlotW + 1);
+				
+			}
+			pCurrent = pTmp;
+		}
+	}
+	
+	virtual int AdjustFontSize(const char* pText, int TextLength, int MaxSize = -1)
+	{
+		int WidthOfText = CalculateTextWidth(pText, TextLength, 0, 100);
+
+		int FontSize = 100.f / ((float)WidthOfText / (float)MaxSize);
 
-					Graphics()->LoadTextureRawSub(TextureID, x + WidthLastChars, y, SlotW, SlotH, CImageInfo::FORMAT_ALPHA, ms_aGlyphData);
-					WidthLastChars += (SlotW + 1);
+		if (MaxSize > 0 && FontSize > MaxSize)
+			FontSize = MaxSize;
+
+		return FontSize;
+	}
+
+	virtual int CalculateTextWidth(const char* pText, int TextLength, int FontWidth, int FontHeight)
+	{
+		CFont* pFont = m_pDefaultFont;
+		const char* pCurrent = (char*)pText;
+		const char* pEnd = pCurrent + TextLength;
+
+		int WidthOfText = 0;
+		FT_Set_Pixel_Sizes(pFont->m_FtFace, FontWidth, FontHeight);
+		while (pCurrent < pEnd)
+		{
+			const char* pTmp = pCurrent;
+			int NextCharacter = str_utf8_decode(&pTmp);
+
+			if (NextCharacter)
```

Comment:
```suggestion
			if(NextCharacter)
```

No space after control keywords